File: Packages/SecuPersoData/Sources/SecuPersoData/MockServices.swift
import Foundation
import SecuPersoDomain

public final class MockLoginActivityService: LoginActivityService, @unchecked Sendable {
    private let coordinator: MockDataCoordinator
    private let streamStore = StreamStore<[LoginEvent]>(initialValue: [])

    public init(coordinator: MockDataCoordinator) {
        self.coordinator = coordinator
    }

    public func refresh() async throws -> [LoginEvent] {
        let events = try await coordinator.refreshLoginEvents()
        streamStore.publish(events)
        return events
    }

    public func stream() -> AsyncStream<[LoginEvent]> {
        streamStore.makeStream()
    }
}

public final class MockIncidentService: IncidentService, IncidentReadableService, @unchecked Sendable {
    private let coordinator: MockDataCoordinator

    public init(coordinator: MockDataCoordinator) {
        self.coordinator = coordinator
    }

    public func create(from loginEventID: UUID) async throws -> IncidentCase {
        try await coordinator.createIncident(from: loginEventID)
    }

    public func resolve(_ incidentID: UUID) async throws {
        try await coordinator.resolveIncident(incidentID)
    }

    public func list() async throws -> [IncidentCase] {
        try await coordinator.listIncidents()
    }
}

public final class MockProviderConnectionService: ProviderConnectionService, ProviderConnectionReadableService, @unchecked Sendable {
    private let coordinator: MockDataCoordinator

    public init(coordinator: MockDataCoordinator) {
        self.coordinator = coordinator
    }

    public func beginConnection(for provider: ProviderID) async -> AsyncStream<ProviderConnectionUpdate> {
        AsyncStream { continuation in
            let task = Task {
                defer { continuation.finish() }

                do {
                    continuation.yield(ProviderConnectionUpdate(state: .connecting, message: "Opening consent screen..."))
                    try await Task.sleep(for: .seconds(1))
                    try Task.checkCancellation()

                    continuation.yield(ProviderConnectionUpdate(state: .connecting, message: "Granting permissions..."))
                    try await Task.sleep(for: .seconds(1))
                    try Task.checkCancellation()

                    let finalState: ConnectionState = provider == .other ? .error : .connected
                    try await coordinator.updateProviderState(provider, state: finalState)
                    let finalMessage = finalState == .connected
                        ? "Provider connected successfully."
                        : "Provider connection failed in mock flow."
                    continuation.yield(ProviderConnectionUpdate(state: finalState, message: finalMessage))
                } catch is CancellationError {
                    return
                } catch {
                    try? await coordinator.updateProviderState(provider, state: .error)
                    continuation.yield(
                        ProviderConnectionUpdate(
                            state: .error,
                            message: "Provider connection failed in mock flow."
                        )
                    )
                }
            }

            continuation.onTermination = { @Sendable _ in
                task.cancel()
            }
        }
    }

    public func disconnect(_ provider: ProviderID) async throws {
        try await coordinator.updateProviderState(provider, state: .disconnected)
    }

    public func connections() async throws -> [ProviderConnection] {
        try await coordinator.listProviderConnections()
    }
}

public final class MockProviderCatalogService: ProviderCatalogService, @unchecked Sendable {
    private let coordinator: MockDataCoordinator

    public init(coordinator: MockDataCoordinator) {
        self.coordinator = coordinator
    }

    public func providers() async throws -> [ProviderDescriptor] {
        try await coordinator.loadProviderCatalog()
    }
}

public final class MockScenarioControlService: ScenarioControlService, @unchecked Sendable {
    private let coordinator: MockDataCoordinator

    public init(coordinator: MockDataCoordinator) {
        self.coordinator = coordinator
    }

    public func setScenario(_ scenario: FixtureScenario) async throws {
        await coordinator.setScenario(scenario)
    }

    public func currentScenario() async -> FixtureScenario {
        await coordinator.currentScenario()
    }
}

public final class MockLoginEventActionService: LoginEventActionService, @unchecked Sendable {
    private let coordinator: MockDataCoordinator

    public init(coordinator: MockDataCoordinator) {
        self.coordinator = coordinator
    }

    public func markAsExpected(_ loginEventID: UUID) async throws -> LoginEvent? {
        try await coordinator.markLoginAsExpected(loginEventID)
    }
}

File: src/data/mockServices.ts
import type {
  ExposureRecord,
  LoginEvent,
  ProviderID,
  ProviderConnection,
  ProviderConnectionUpdate,
  ProviderDescriptor,
  IncidentCase,
  FixtureScenario,
  MonitoredEmailAddress,
} from '@domain/models';
import { PROVIDER_IDS, providerDisplayName } from '@domain/models';
import type {
  ExposureMonitoringService,
  LoginActivityService,
  ProviderConnectionService,
  IncidentService,
  IncidentReadableService,
  ProviderCatalogService,
  ProviderConnectionReadableService,
  ScenarioControlService,
  LoginEventActionService,
  MonitoredEmailService,
  ExposureSourceConfigurationService,
} from '@domain/protocols';
import { StreamStore } from './streamStore';

export class MockLoginActivityService implements LoginActivityService {
  private coordinator: import('./mockDataCoordinator').MockDataCoordinator;
  private streamStore = new StreamStore<LoginEvent[]>([]);

  constructor(coordinator: import('./mockDataCoordinator').MockDataCoordinator) {
    this.coordinator = coordinator;
  }

  async refresh(): Promise<LoginEvent[]> {
    const events = await this.coordinator.refreshLoginEvents();
    this.streamStore.publish(events);
    return events;
  }

  stream(callback: (values: LoginEvent[]) => void): () => void {
    return this.streamStore.subscribe(callback);
  }
}

export class MockIncidentService implements IncidentService, IncidentReadableService {
  private coordinator: import('./mockDataCoordinator').MockDataCoordinator;

  constructor(coordinator: import('./mockDataCoordinator').MockDataCoordinator) {
    this.coordinator = coordinator;
  }

  async create(loginEventID: string): Promise<IncidentCase> {
    return this.coordinator.createIncident(loginEventID);
  }

  async resolve(incidentID: string): Promise<void> {
    return this.coordinator.resolveIncident(incidentID);
  }

  async list(): Promise<IncidentCase[]> {
    return this.coordinator.listIncidents();
  }
}

export class MockProviderConnectionService implements ProviderConnectionService, ProviderConnectionReadableService {
  private coordinator: import('./mockDataCoordinator').MockDataCoordinator;

  constructor(coordinator: import('./mockDataCoordinator').MockDataCoordinator) {
    this.coordinator = coordinator;
  }

  async beginConnection(provider: ProviderID, onUpdate: (update: ProviderConnectionUpdate) => void): Promise<() => void> {
    const cancelled = { value: false };

    const run = async () => {
      onUpdate({ state: 'connecting', message: 'Opening consent screen...' });
      await new Promise((resolve) => setTimeout(resolve, 1000));
      if (cancelled.value) return;

      onUpdate({ state: 'connecting', message: 'Granting permissions...' });
      await new Promise((resolve) => setTimeout(resolve, 1000));
      if (cancelled.value) return;

      const finalState = provider === 'other' ? 'error' : 'connected';
      await this.coordinator.updateProviderState(provider, finalState);
      const finalMessage = finalState === 'connected'
        ? 'Provider connected successfully.'
        : 'Provider connection failed in mock flow.';
      onUpdate({ state: finalState, message: finalMessage });
    };

    run();
    return () => { cancelled.value = true; };
  }

  async disconnect(provider: ProviderID): Promise<void> {
    return this.coordinator.updateProviderState(provider, 'disconnected');
  }

  async connections(): Promise<ProviderConnection[]> {
    return this.coordinator.listProviderConnections();
  }
}

export class MockProviderCatalogService implements ProviderCatalogService {
  private coordinator: import('./mockDataCoordinator').MockDataCoordinator;

  constructor(coordinator: import('./mockDataCoordinator').MockDataCoordinator) {
    this.coordinator = coordinator;
  }

  async providers(): Promise<ProviderDescriptor[]> {
    return this.coordinator.loadProviderCatalog();
  }
}

export class MockScenarioControlService implements ScenarioControlService {
  private coordinator: import('./mockDataCoordinator').MockDataCoordinator;

  constructor(coordinator: import('./mockDataCoordinator').MockDataCoordinator) {
    this.coordinator = coordinator;
  }

  async setScenario(scenario: FixtureScenario): Promise<void> {
    this.coordinator.setScenario(scenario);
  }

  async currentScenario(): Promise<FixtureScenario> {
    return this.coordinator.currentScenario();
  }
}

export class MockLoginEventActionService implements LoginEventActionService {
  private coordinator: import('./mockDataCoordinator').MockDataCoordinator;

  constructor(coordinator: import('./mockDataCoordinator').MockDataCoordinator) {
    this.coordinator = coordinator;
  }

  async markAsExpected(loginEventID: string): Promise<LoginEvent | null> {
    return this.coordinator.markLoginAsExpected(loginEventID);
  }
}